File: daragacivilregistrymanagementsystem/static/js/document_annotation.js
(function () {
  function attr(root, name) {
    return (root.getAttribute(name) || "").trim();
  }

  function fieldVal(scope, name) {
    var el = scope.querySelector('[name="' + name + '"]');
    return el && el.value ? String(el.value).trim() : "";
  }

  function buildLegitimation(root) {
    var form = root.closest("form") || document;
    var father = fieldVal(form, "Name of Father") || attr(root, "data-father");
    var mother = fieldVal(form, "Name of Mother") || attr(root, "data-mother");
    var child =
      fieldVal(root, "annotation_new_name") ||
      fieldVal(form, "Name of Child") ||
      attr(root, "data-child");
    var marriageDate =
      fieldVal(form, "Date of Marriage of Parents") ||
      fieldVal(form, "Date of Marriage") ||
      attr(root, "data-marriage-date");
    var marriagePlace =
      fieldVal(form, "Place of Marriage of Parents") ||
      fieldVal(form, "Place of Marriage") ||
      attr(root, "data-marriage-place");
    var registry = fieldVal(root, "annotation_marriage_registry");

    var parts = ["LEGITIMATED BY SUBSEQUENT MARRIAGE OF PARENTS"];
    if (father && mother) parts.push(father + " AND " + mother);
    else if (father || mother) parts.push(father || mother);
    if (marriageDate) parts.push("ON " + marriageDate);
    if (marriagePlace) parts.push("AT " + marriagePlace);
    if (registry) parts.push("UNDER REGISTRY NO. " + registry);
    var sentence = parts.join(" ");
    if (child) sentence += ", THE CHILD SHALL BE KNOWN AS " + child;
    sentence = sentence.replace(/\s+/g, " ").trim();
    if (sentence && sentence.charAt(sentence.length - 1) !== ".") sentence += ".";
    return sentence.toUpperCase();
  }

  function syncStamps(root) {
    var textEl = root.querySelector('[name="annotation_text"]');
    var signEl = root.querySelector('[name="annotation_signatory"]');
    var titleEl = root.querySelector('[name="annotation_title"]');
    var text = textEl && textEl.value ? String(textEl.value).trim() : "";
    var signatory = signEl && signEl.value ? String(signEl.value).trim() : "";
    var title = titleEl && titleEl.value ? String(titleEl.value).trim() : "";
    var stamps = document.querySelectorAll("[data-annotation-stamp]");
    for (var i = 0; i < stamps.length; i += 1) {
      var stamp = stamps[i];
      var textNode = stamp.querySelector(".doc-annotation-text");
      var signNode = stamp.querySelector(".doc-annotation-signatory");
      var titleNode = stamp.querySelector(".doc-annotation-title");
      if (textNode) textNode.textContent = text;
      if (signNode) signNode.textContent = signatory;
      if (titleNode) titleNode.textContent = title;
      if (text) stamp.classList.remove("is-empty");
      else stamp.classList.add("is-empty");
    }
  }

  function bind(root) {
    var fillBtn = root.querySelector("[data-annotation-fill]");
    var textEl = root.querySelector('[name="annotation_text"]');
    if (fillBtn && textEl) {
      fillBtn.addEventListener("click", function () {
        textEl.value = buildLegitimation(root);
        syncStamps(root);
      });
    }
    var inputs = root.querySelectorAll(
      '[name="annotation_text"], [name="annotation_signatory"], [name="annotation_title"]'
    );
    for (var i = 0; i < inputs.length; i += 1) {
      inputs[i].addEventListener("input", function () {
        syncStamps(root);
      });
    }
  }

  document.querySelectorAll("[data-annotation-form]").forEach(bind);
})();

File: daragacivilregistrymanagementsystem/services/document_annotation.py
"""Official certificate annotation (typed LCR box + registrar signature)."""

from __future__ import annotations

from typing import Any

STORAGE_KEY = "_annotation"

KINDS = (
    ("legitimation", "Legitimation by subsequent marriage"),
    ("adoption", "Adoption"),
    ("court_order", "Court order / decree"),
    ("correction", "Correction of entry"),
    ("other", "Other annotation"),
)

DEFAULT_SIGNATORY = "MICHELLE M. MARINAY"
DEFAULT_TITLE = "MGDH I / MUN. CIVIL REGISTRAR"

FORM_KEYS = frozenset(
    {
        "annotation_kind",
        "annotation_text",
        "annotation_signatory",
        "annotation_title",
        "annotation_new_name",
        "annotation_marriage_registry",
    }
)


def public_fields(data: dict | None) -> dict:
    if not isinstance(data, dict):
        return {}
    return {k: v for k, v in data.items() if not str(k).startswith("_")}


def from_data(data: dict | None) -> dict | None:
    if not isinstance(data, dict):
        return None
    raw = data.get(STORAGE_KEY)
    if not isinstance(raw, dict):
        return None
    text = str(raw.get("text") or "").strip()
    if not text:
        return None
    return {
        "kind": str(raw.get("kind") or "other").strip() or "other",
        "text": text,
        "signatory": str(raw.get("signatory") or "").strip() or DEFAULT_SIGNATORY,
        "title": str(raw.get("title") or "").strip() or DEFAULT_TITLE,
        "new_name": str(raw.get("new_name") or "").strip(),
        "marriage_registry": str(raw.get("marriage_registry") or "").strip(),
    }


def form_defaults(data: dict | None) -> dict:
    ann = from_data(data) or {}
    return {
        "kind": ann.get("kind") or "legitimation",
        "text": ann.get("text") or "",
        "signatory": ann.get("signatory") or DEFAULT_SIGNATORY,
        "title": ann.get("title") or DEFAULT_TITLE,
        "new_name": ann.get("new_name") or "",
        "marriage_registry": ann.get("marriage_registry") or "",
    }


def from_form(form) -> dict | None:
    text = (form.get("annotation_text") or "").strip()
    if not text:
        return None
    return {
        "kind": (form.get("annotation_kind") or "other").strip() or "other",
        "text": text,
        "signatory": (form.get("annotation_signatory") or "").strip() or DEFAULT_SIGNATORY,
        "title": (form.get("annotation_title") or "").strip() or DEFAULT_TITLE,
        "new_name": (form.get("annotation_new_name") or "").strip(),
        "marriage_registry": (form.get("annotation_marriage_registry") or "").strip(),
    }


def apply_to_data(data: dict | None, annotation: dict | None) -> dict:
    out = dict(data or {})
    out.pop(STORAGE_KEY, None)
    for key in list(out.keys()):
        if key in FORM_KEYS or str(key).startswith("annotation_"):
            out.pop(key, None)
    if annotation:
        out[STORAGE_KEY] = annotation
    return out


def build_legitimation_text(
    data: dict | None,
    *,
    new_name: str = "",
    marriage_registry: str = "",
) -> str:
    src: dict[str, Any] = data if isinstance(data, dict) else {}
    father = str(src.get("Name of Father") or "").strip()
    mother = str(src.get("Name of Mother") or "").strip()
    marriage_date = str(
        src.get("Date of Marriage of Parents")
        or src.get("Date of marriage of Parents")
        or src.get("Date of Marriage")
        or ""
    ).strip()
    marriage_place = str(
        src.get("Place of Marriage of Parents")
        or src.get("Place of marriage of Parents")
        or src.get("Place of Marriage")
        or ""
    ).strip()
    child = str(new_name or src.get("Name of Child") or "").strip()
    registry = str(marriage_registry or "").strip()

    parts = ["LEGITIMATED BY SUBSEQUENT MARRIAGE OF PARENTS"]
    if father and mother:
        parts.append(f"{father} AND {mother}")
    elif father or mother:
        parts.append(father or mother)
    if marriage_date:
        parts.append(f"ON {marriage_date}")
    if marriage_place:
        parts.append(f"AT {marriage_place}")
    if registry:
        parts.append(f"UNDER REGISTRY NO. {registry}")
    sentence = " ".join(parts)
    if child:
        sentence += f", THE CHILD SHALL BE KNOWN AS {child}"
    sentence = sentence.strip()
    if not sentence.endswith("."):
        sentence += "."
    return sentence.upper()
